arch: Allow to specify memory for S2RAM resume

tomchy opened the pull request:
If the area, dedicated for the interrupt stack is not available, allow to specify a memory region that will be used as the stack for the S2RAM resume logic.

tomchy commented:
> > If the area, dedicated for the interrupt stack is not available,
> 
> Could you provide more details?

The current solution assumes that the value of `z_interrupt_stacks` is known, which might not be the case if:
 - The project uses bootloader (i.e. MCUboot).
 - The MCUboot chainloads the application.
 - The application enters S2RAM.
 - The SoC resumes from S2RAM from reset vector (due to the boot order - inside MCUboot).
 - The MCUboot must detect and jump to the application code to resume from S2RAM.
 - The MCUboot code will execute the problematic `ldr r0` as well as `msr msp, r0` instructions, possibly corrupting application data, before checking the S2RAM flags.

mathieuchopstm reviewed (changes requested):
There needs to be documentation somewhere about these special regions or it will get out of hand. While at it, please document `pm_s2ram_context` too - I think it was introduced without docs either.

tomchy commented:
> > There needs to be documentation somewhere about these special regions or it will get out of hand. While at it, please document `pm_s2ram_context` too - I think it was introduced without docs either.
> 
> Do you have any specific area in the Zephyr documentation that you'd find suitable for such page?

Extended the `PM_S2RAM` Kconfig symbol help message (this is the same place that mentions the `pm_s2ram` section).

mathieuchopstm commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:
Can you split this change to a separate commit?

tomchy committed:
arch: Allow to specify memory for S2RAM resume

If the area, dedicated for the interrupt stack is not available, allow
to specify a memory region that will be used as the stack for the S2RAM
resume logic.

Signed-off-by: Tomasz Chyrowicz <tomasz.chyrowicz@nordicsemi.no>

tomchy committed:
boards: Define pm_s2ram_stack for nRF54H20

Add the definition of pm_s2ram_stack memory region for nRF54H20.

Signed-off-by: Tomasz Chyrowicz <tomasz.chyrowicz@nordicsemi.no>

mathieuchopstm commented on arch/arm/core/cortex_m/reset.S:
```diff
@@ -113,6 +124,7 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
      * a short while, there is no change in behavior in either of the paths.
      */
     ldr r0, =z_interrupt_stacks + CONFIG_ISR_STACK_SIZE + MPU_GUARD_ALIGN_AND_SIZE
+#endif
```
I'd still invert both sections such that `ldr r0, =z_interrupt_stacks` comes before?
i.e.,
```diff
#if defined(CONFIG_PM_S2RAM)
+#if !DT_NODE_EXISTS(DT_NODELABEL(pm_s2ram_stack)) ||\
+	!DT_NODE_HAS_COMPAT(DT_NODELABEL(pm_s2ram_stack), zephyr_memory_region)
    /*
     * Temporarily set MSP to interrupt stack so that arch_pm_s2ram_resume can
     * use stack for calling pm_s2ram_mark_check_and_clear.
     * This is safe because suspend never being called from an ISR ensures that
     * interrupt stack was not live during suspend.
     * Note:
     * if resuming from suspend, MSP is restored from cpu context
     * if not resuming, MSP will be set back to z_main_stack for proper init
     * And, apart from arch_pm_s2ram_resume being able to use the stack for
     * a short while, there is no change in behavior in either of the paths.
     */
    ldr r0, =z_interrupt_stacks + CONFIG_ISR_STACK_SIZE + MPU_GUARD_ALIGN_AND_SIZE
+#else
+    /* In certain scenarios, the interrupt stack is actually not safe to overwrite.
+     * For example, when MCUboot is used, the bootloader's "z_interrupt_stack" and the
+     * loaded image's "z_interrupt_stacks" are NOT at the same address, and writing to
+     * the former would corrupt unrelated data from the loaded image. To support such
+     * scenarios, if the Device Tree provides a specially named "zephyr,memory-region",
+     * use it as the stack to run arch_pm_s2ram_resume instead of the interrupt stack.
+     */
+    ldr r0, =DT_REG_ADDR(DT_NODELABEL(pm_s2ram_stack)) + DT_REG_SIZE(DT_NODELABEL(pm_s2ram_stack))
+#endif
```

mathieuchopstm replied:
so let's leave it as-is

mathieuchopstm reviewed (dismissed):
non-blocker:

mathieuchopstm commented:
> The flow @mathieuchopstm mentioned looks similar to what was mentioned in [#96290 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/96290#discussion_r2363161498). If that's the case then maybe a new `soc_resume_hook` api would be better solution:
> 
> ```
> #if defined(CONFIG_PM_S2RAM)
> #if defined(CONFIG_SOC_RESUME_HOOK)
>     bl soc_resume_hook
> #else
>     /*
>      * Temporarily set MSP to interrupt stack so that arch_pm_s2ram_resume can
>      * use stack for calling pm_s2ram_mark_check_and_clear.
> 	@@ -113,6 +124,7 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
>      * a short while, there is no change in behavior in either of the paths.
>      */
>     ldr r0, =z_interrupt_stacks + CONFIG_ISR_STACK_SIZE + MPU_GUARD_ALIGN_AND_SIZE
> #endif
> #endif
> ```
> 
> and `soc_resume_hook` does what they expect to be done in `arch_pm_s2ram_resume()` similar to https://github.com/zephyrproject-rtos/zephyr/pull/96290/files#diff-15c65cbae35c06ba0b60c2a5129e9da316484d76b96dc60b689af15d54996601R19.

#96290's approach has the advantage that it _runs through the bootloader again_ which may be required in certain use-cases (e.g., XIP from external flash/RAM?); the disadvantage is that it runs through the bootloader again which may be inappropriate due to shared resource overwriting... but it is gated behind a Kconfig.

tomchy commented:
@wearyzen, @mathieuchopstm You're correct. This change alone does not solve the problem of resuming from S2RAM if bootloader is enabled - it must be combined with one of the other changes:
1. https://github.com/zephyrproject-rtos/zephyr/pull/93818: to share the S2RAM context between the application and bootloader. That way the S2RAM routine executed from the bootloader code resumes the state created by the application. This solution however creates a restriction, that the resume procedure must be exactly the same in both the bootloader and all application builds, including all possible updates. That's why a new method is being proposed (see pt 2).
2. https://github.com/zephyrproject-rtos/zephyr/pull/96290: to execute a small, PM-independent piece of logic, just so the bootloader can detect which variant of the application (Direct XIP) is the active one and jump to a reset vector of that application. That way the S2RAM resume routine from the application is handling the context and restores the CPU state.

@wearyzen: regarding the custom stack, I agree that we may name this "small" stack a `custom_init_stack`. Do you think adding this definition in this PR is enough or do you envision some other functional changes to be done here?

tomchy commented:
> > @wearyzen: regarding the custom stack, I agree that we may name this "small" stack a `custom_init_stack`. Do you think adding this definition in this PR is enough or do you envision some other functional changes to be done here?
> 
> Considering both the PRs, I think changing MSP to use to small stack can just be done inside the soc_resume_hook() and there is no change needed here at arch level.

Do you propose to create a new `SOC_RESUME_HOOK` and call it instead of the `ldr` and `msr` (and possibly `bl arch_pm_s2ram_resume`) instructions of arch-level `__start` implementation?

I cannot find a reference for the existing `soc_resume_hook()` function.

wearyzen commented:
> Do you propose to create a new `SOC_RESUME_HOOK` and call it instead of the `ldr` and `msr` (and possibly `bl arch_pm_s2ram_resume`) instructions of arch-level `__start` implementation?

Yes a new hook suggested to be added in the other PR (https://github.com/zephyrproject-rtos/zephyr/pull/96290#issuecomment-3333319836)

wearyzen commented:
> Regarding the new hook - note that the `soc_resume_hook` cannot `push` anything to the stack before it sets the `msp`. With some additional care to the `lr` it should be fixable though.
> 
> What I'm more concerned about - the issue is not specific to the SoC, but rather to the combination of having both bootloader and a sleepy application on the same core, started through chain-loading. That said, are you sure that a fix for it should be hidden inside nRF-specific files?
> 
> This kind of issues are not easy to debug, so before we remove it from the generic scope, I would like to confirm that the solution on other platforms is not desired.

I agree that there would be an issue is such a scenario and might not be easy to debug but for such a design where mcuboot resumes just to load the next images context won't they still need the soc_resume_hook? and if they do then with the right documentation they may chose to either change the msp to something like this PR does or not use the the stack at all and avoid the issue in soc_resume_hook 🤷.
But I agree that with `soc_resume_hook` added there should be good documentation highlighting this issue.

tomchy commented:
> Please rebase and resolve conflicts. In this PR https://github.com/zephyrproject-rtos/zephyr/pull/95474/files size of `cpuapp_ram0` was decreased by 224 B (256-32) while only 192 B were assigned to `pm_ramfunc`. You should be able to fit in that gap.

PR rebased, docs removed due to: https://github.com/zephyrproject-rtos/zephyr/commit/4626f6b776f662702934ac85929edc35dce832b1

To my understanding - if possible, the solution should be move into the board-specific files and there is an agreement on adding a platform-specific hook.
